compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -121,15 +122,31 @@ impl<Tag> Allocation<Tag> {
         Allocation::from_bytes(slice, Align::ONE, Mutability::Not)
     }
 
-    pub fn uninit(size: Size, align: Align) -> Self {
-        Allocation {
-            bytes: vec![0; size.bytes_usize()],
+    /// Try to create an Allocation of `size` bytes, failing if there is not enough memory
+    /// available to the compiler to do so.
+    pub fn uninit(size: Size, align: Align) -> InterpResult<'static, Self> {
+        let mut bytes = Vec::new();
+        bytes.try_reserve(size.bytes_usize()).map_err(|_| {
+            // This results in an error that can happen non-deterministically, since the memory
+            // available to the compiler can change between runs. Normally queries are always
+            // deterministic. However, we can be non-determinstic here because all uses of const
+            // evaluation do one of:
+            // - cause a fatal compiler error when they see this error as the result of const
+            //   evaluation
+            // - panic on evaluation failure
+            // - always evaluate very small constants that are practically unlikely to result in
+            //   memory exhaustion
```

Comment:
I don't get the last bullet point. Why do we even need to make a probabilistic argument?

You can ignore the cases where callers panic on alloc failure; that is the same as before. So I think the argument is simply: all users of const-evaluation will make the compilation fail upon encountering `MemoryExhausted`.

@oli-obk is that true? I seem to recall ConstProp ignoring many kinds of errors...

Reply:
I've updated the comment to be more detailed, I _think_ this covers all of the cases now. I changed `ConstProp` so it always hard errors on allocation failure to prevent it from optimizing differently based on the environment.

Reply:
Right, fixing any particular optimization is not that hard -- but is ConstProp the only one that needs fixing? Can we somehow ensure we got all of them?

Reply:
I don't think being more detailed is necessarily good here, that makes it more prone to being outdated and it makes it harder to grasp the actual important point: *all users of const-evaluation will make the compilation fail upon encountering `MemoryExhausted`*.

Reply:
I believe that ConstProp is the only optimization pass that uses the Miri engine to evaluate constants directly: as far as I'm aware, the other passes rely on ConstProp to have the constants already propagated in the MIR for them.

Reply:
Yeah that sounds plausible... ConstProp has its own `machine` instance even, everything else should use the queries (and thus hard error).

Reply:
It turns out I missed a few instances of const-evaluation in ConstProp, I just added checks for those